be/src/exprs/agg/distinct.h

```diff
@@ -249,23 +271,56 @@ class TDistinctAggregateFunction
         size_t mem_usage = 0;
         auto& agg_state = this->data(state);
 
-        if constexpr (IsSlice<T>) {
-            MemPool* mem_pool = ctx->impl()->mem_pool();
-            for (size_t i = 0; i < batch_size; ++i) {
-                mem_usage += agg_state.update(mem_pool, column->get_slice(i));
+        struct CacheEntry {
+            size_t hash_value;
+        };
+
+        std::vector<CacheEntry> cache(batch_size);
+        const auto& container_data = column->get_data();
+        for (size_t i = 0; i < batch_size; ++i) {
+            size_t hash_value = agg_state.set.hash_function()(container_data[i]);
+            cache[i] = CacheEntry{hash_value};
+        }
+        size_t prefetch_index = 16;
+
+        MemPool* mem_pool = ctx->impl()->mem_pool();
+        for (size_t i = 0; i < batch_size; ++i) {
+            if (prefetch_index < batch_size) {
+                agg_state.set.prefetch_hash(cache[prefetch_index].hash_value);
+                prefetch_index++;
             }
-        } else {
-            auto* data = column->get_data().data();
-            // after doing some benchmark, we see a lot of perf gain by prefetching when doing count(distinct)
-            // but only when there is no group by clause or group by number is low.
-            // prefetching has down side if group by number is high.
-            static const size_t prefetch_dist = 4;
-            for (size_t i = 0; i < batch_size; ++i) {
-                if ((i + prefetch_dist) < batch_size) {
-                    agg_state.prefetch(data[i + prefetch_dist]);
-                }
-                mem_usage += agg_state.update(data[i]);
+            mem_usage += agg_state.update_with_hash(mem_pool, container_data[i], cache[i].hash_value);
+        }
+        ctx->impl()->add_mem_usage(mem_usage);
+    }
+
+    void update_batch(FunctionContext* ctx, size_t batch_size, size_t state_offset, const Column** columns,
```

Comment:
Better to add comment to explain why split this update operation into two loop.
And why use prefetch.

Reply:
comments added.